File: sdk/python/alp_sdk/macro.py
"""ALP MacroEngine — Dynamic @macro object generation (v37.0.0 — Python SDK parity).

Expands `@macro` blocks into concrete protocol objects by evaluating `iterate_over`
expressions and interpolating `${...}` template variables.
"""
from __future__ import annotations

import json
import re
from typing import Any, Dict, List, Optional, Set

from .alpel import evaluate


class MacroDefinition:
    def __init__(self, id: str, iterate_over: str, template: Dict[str, Any], name: Optional[str] = None, as_: str = "item"):
        self.id = id
        self.iterate_over = iterate_over
        self.template = template
        self.name = name
        self.var_name = as_


def _interpolate_string(template: str, var_name: str, item: Any) -> str:
    def replacer(match: re.Match) -> str:
        expr = match.group(1).strip()
        if expr == var_name:
            return json.dumps(item) if isinstance(item, (dict, list)) else str(item)
        if expr.startswith(var_name + "."):
            prop = expr[len(var_name) + 1:]
            parts = prop.split(".")
            val = item
            for p in parts:
                if val is None or not isinstance(val, dict):
                    return ""
                val = val.get(p)
            return str(val) if val is not None else ""
        return match.group(0)

    return re.sub(r"\$\{([^}]+)\}", replacer, template)


def _interpolate_deep(obj: Any, var_name: str, item: Any) -> Any:
    if isinstance(obj, str):
        return _interpolate_string(obj, var_name, item)
    if isinstance(obj, list):
        return [_interpolate_deep(v, var_name, item) for v in obj]
    if isinstance(obj, dict):
        return {
            _interpolate_string(k, var_name, item): _interpolate_deep(v, var_name, item)
            for k, v in obj.items()
        }
    return obj


class MacroEngine:
    def expand(self, macro: Dict[str, Any], context: Optional[Dict[str, Any]] = None) -> List[Dict[str, Any]]:
        macro_id = macro.get("id", "unnamed")
        iterate_over = macro.get("iterate_over")
        template = macro.get("template")
        var_name = macro.get("as", "item")

        if not iterate_over:
            raise ValueError(f"Macro '{macro_id}': missing iterate_over")
        if not template:
            raise ValueError(f"Macro '{macro_id}': missing template")

        raw = str(iterate_over).strip()
        if raw.startswith("["):
            # Try parsing JSON array directly
            try:
                items = json.loads(raw.replace("'", '"'))
            except Exception:
                items = evaluate(raw, context or {})
        else:
            items = evaluate(raw, context or {})

        if not isinstance(items, list):
            items = [items]

        expanded: List[Dict[str, Any]] = []
        seen_ids: Set[str] = set()

        for item in items:
            obj = _interpolate_deep(template, var_name, item)
            if isinstance(obj, dict):
                obj["_sourceMacro"] = macro_id
                obj_id = obj.get("id")
                if obj_id:
                    if obj_id in seen_ids:
                        raise ValueError(f"Macro '{macro_id}': duplicate generated id '{obj_id}'")
                    seen_ids.add(obj_id)
            expanded.append(obj)

        return expanded

    def expand_all(self, objects: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        result: List[Dict[str, Any]] = []
        for obj in objects:
            if isinstance(obj, dict) and obj.get("_type") == "macro" and "iterate_over" in obj and "template" in obj:
                expanded = self.expand(obj)
                result.extend(expanded)
            else:
                result.append(obj)
        return result

File: parser/src/macro.ts
/**
 * ALP MacroEngine — Dynamic @macro object generation (v37.0.0).
 *
 * Expands `@macro` blocks into concrete protocol objects by evaluating
 * the `iterate_over` ALPEL expression and interpolating `${...}` template
 * variables in the `template` block for each iteration item.
 *
 * Spec reference: spec/03-protocol-objects.md §20.
 */

import { evaluate } from './alpel';

export interface MacroDefinition {
  id: string;
  name?: string;
  iterate_over: string;
  as?: string;
  template: Record<string, any>;
}

export interface ExpandedObject {
  _type: string;
  id: string;
  _sourceMarco?: string;
  [key: string]: any;
}

/**
 * Interpolate `${var}` and `${var.prop}` placeholders in a string.
 */
function interpolateString(template: string, varName: string, item: any): string {
  return template.replace(/\$\{([^}]+)\}/g, (_match, expr: string) => {
    const trimmed = expr.trim();
    if (trimmed === varName) {
      return typeof item === 'object' ? JSON.stringify(item) : String(item);
    }
    if (trimmed.startsWith(varName + '.')) {
      const prop = trimmed.slice(varName.length + 1);
      const parts = prop.split('.');
      let val: any = item;
      for (const p of parts) {
        if (val == null) return '';
        val = val[p];
      }
      return val != null ? String(val) : '';
    }
    return _match;
  });
}

/**
 * Recursively interpolate all string values in an object/array.
 */
function interpolateDeep(obj: any, varName: string, item: any): any {
  if (typeof obj === 'string') {
    return interpolateString(obj, varName, item);
  }
  if (Array.isArray(obj)) {
    return obj.map(v => interpolateDeep(v, varName, item));
  }
  if (obj && typeof obj === 'object') {
    const result: Record<string, any> = {};
    for (const [k, v] of Object.entries(obj)) {
      result[interpolateString(k, varName, item)] = interpolateDeep(v, varName, item);
    }
    return result;
  }
  return obj;
}

export class MacroEngine {
  /**
   * Expand a single `@macro` definition into concrete objects.
   */
  expand(macro: MacroDefinition, context?: Record<string, any>): ExpandedObject[] {
    if (!macro.iterate_over) {
      throw new Error(`Macro '${macro.id}': missing iterate_over`);
    }
    if (!macro.template) {
      throw new Error(`Macro '${macro.id}': missing template`);
    }

    // Evaluate the iterate_over expression
    let items: any[];
    try {
      const raw = macro.iterate_over.trim();
      // Try JSON array first
      if (raw.startsWith('[')) {
        items = JSON.parse(raw.replace(/'/g, '"'));
      } else {
        // Evaluate as ALPEL expression
        const result = evaluate(raw, context || {});
        items = Array.isArray(result) ? result : [result];
      }
    } catch {
      throw new Error(`Macro '${macro.id}': failed to evaluate iterate_over: ${macro.iterate_over}`);
    }

    if (!Array.isArray(items)) {
      throw new Error(`Macro '${macro.id}': iterate_over must resolve to an array`);
    }

    const varName = macro.as || 'item';
    const expanded: ExpandedObject[] = [];
    const seenIds = new Set<string>();

    for (const item of items) {
      const obj = interpolateDeep(macro.template, varName, item);
      obj._sourceMacro = macro.id;

      if (obj.id && seenIds.has(obj.id)) {
        throw new Error(`Macro '${macro.id}': duplicate generated id '${obj.id}'`);
      }
      if (obj.id) seenIds.add(obj.id);

      expanded.push(obj as ExpandedObject);
    }

    return expanded;
  }

  /**
   * Scan a list of parsed objects, expand all `@macro` entries,
   * and return a flat list with macros replaced by their expansions.
   */
  expandAll(objects: any[]): any[] {
    const result: any[] = [];
    for (const obj of objects) {
      if (obj._type === 'macro' && obj.iterate_over && obj.template) {
        const expanded = this.expand(obj as MacroDefinition);
        result.push(...expanded);
      } else {
        result.push(obj);
      }
    }
    return result;
  }
}
